File: member/mall-api/src/main/java/com/laf/mall/api/repository/RaffleRepository.java
package com.laf.mall.api.repository;

import com.laf.mall.api.dto.RaffleInfo;
import com.laf.mall.api.dto.RaffleLog;
import com.laf.mall.api.dto.RaffleReward;
import com.laf.mall.api.querycondition.raffle.MyRewardQueryCondition;
import com.laf.mall.api.utils.datetime.DateTimeUtils;
import com.laf.mall.api.vo.MyRewardInfo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataAccessException;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.PreparedStatementCreator;
import org.springframework.jdbc.core.ResultSetExtractor;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.rowset.SqlRowSet;
import org.springframework.stereotype.Repository;

import java.sql.*;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Repository
public class RaffleRepository {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private DateTimeUtils dateTimeUtils;

    public RaffleInfo selectRaffleInfoByActivityCode(final String activityCode, final Integer mallId) {

        String sql = "SELECT raffle_id,title,raffle_time_start,raffle_time_end,required_points,activity_code,picture,content " +
                "FROM `T_RAFFLE_INFO` ri LEFT JOIN `T_ARTICLES` r ON intro=article_id " +
                "WHERE ri.mall_id=? AND activity_code=? AND raffle_time_start<=? AND raffle_time_end>=?";

        Long curr = new Date().getTime();

        List<RaffleInfo> list = jdbcTemplate.query(sql, new Object[]{mallId, activityCode, curr, curr}, new ResultSetExtractor<List<RaffleInfo>>() {
            @Override
            public List<RaffleInfo> extractData(ResultSet rs) throws SQLException, DataAccessException {
                List<RaffleInfo> _list = new ArrayList<>();

                while (rs.next()) {
                    RaffleInfo _info = new RaffleInfo();
                    _info.setRaffle_id(rs.getInt("raffle_id"));
                    _info.setTitle(rs.getString("title"));
                    _info.setRaffle_time_start(new Date(rs.getLong("raffle_time_start")));
                    _info.setRaffle_time_end(new Date(rs.getLong("raffle_time_end")));
                    _info.setRequired_points(rs.getInt("required_points"));
                    _info.setActivity_code(rs.getString("activity_code"));
                    _info.setPicture(rs.getString("picture"));
                    _info.setIntro(rs.getString("content"));
                    List<RaffleReward> list = selectRaffleRewardList(_info.getRaffle_id());
                    _info.setRewards(list);
                    _list.add(_info);
                }
                return _list;
            }
        });

        return list.isEmpty() ? null : list.get(0);
    }

    public int insertRaffleLog(final RaffleLog log) {
        String sql = " INSERT INTO `T_RAFFLE_LOG` (raffle_id,member_id,member_name,member_mobile,trr_id,action_time) VALUES (?,?,?,?,?,?)";

        GeneratedKeyHolder holder = new GeneratedKeyHolder();
        jdbcTemplate.update(new PreparedStatementCreator() {
            @Override
            public PreparedStatement createPreparedStatement(Connection con) throws SQLException {
                PreparedStatement ps = con.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
                ps.setInt(1, log.getRaffle_id());
                ps.setInt(2, log.getMember_id());
                ps.setString(3, log.getMember_name());
                ps.setString(4, log.getMember_mobile());
                ps.setInt(5, log.getTrr_id());
                ps.setLong(6, log.getAction_time());
                return ps;
            }
        }, holder);

        return holder.getKey().intValue();
    }

    public List<MyRewardInfo> selectMyRewardList(MyRewardQueryCondition condition) {
        String sql = "SELECT action_time,reward_type,reward_value,coupon_name,price,reward_picture FROM `T_RAFFLE_LOG` rl " +
                "INNER JOIN `T_RAFFLE_REWARD` rr ON rl.trr_id=rr.trr_id " +
                "LEFT JOIN `T_COUPON_INFO` ci ON rr.reward_value=ci.coupon_id " +
                "WHERE rl.member_id=?";

        StringBuilder sb = new StringBuilder(sql);
        ArrayList<Object> args = new ArrayList<>();
        ArrayList<Integer> argsType = new ArrayList<>();

        args.add(condition.getMemberId());
        argsType.add(Types.INTEGER);

        sb.append(" ORDER BY action_time DESC LIMIT ?,?");
        args.add(condition.getOffset());
        args.add(condition.getSize());
        argsType.add(Types.INTEGER);
        argsType.add(Types.INTEGER);

        int[] arr = new int[argsType.size()];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = argsType.get(i);
        }

        List<MyRewardInfo> list = jdbcTemplate.query(sb.toString(), args.toArray(), arr, new ResultSetExtractor<List<MyRewardInfo>>() {
            @Override
            public List<MyRewardInfo> extractData(ResultSet rs) throws SQLException, DataAccessException {
                List<MyRewardInfo> _list = new ArrayList<>();

                while (rs.next()) {
                    MyRewardInfo info = new MyRewardInfo();
                    info.setAction_time(new Date(rs.getLong("action_time")));
                    info.setReward_type(rs.getInt("reward_type"));
                    info.setReward_value(rs.getInt("reward_value"));
                    info.setCoupon_name(rs.getString("coupon_name"));
                    info.setPrice(rs.getBigDecimal("price"));
                    info.setReward_picture(rs.getString("reward_picture"));

                    _list.add(info);
                }
                return _list;
            }
        });

        return list;
    }

    public List<RaffleReward> selectRaffleRewardList(final int raffleId) {

        String sql = "SELECT trr_id,raffle_id,reward_name,reward_type,reward_picture,inventory,hit_rate,reward_value,sort_id " +
                "FROM `T_RAFFLE_REWARD` WHERE raffle_id=? ORDER BY sort_id ASC";

        List<RaffleReward> list = jdbcTemplate.query(sql, new Object[]{raffleId}, new ResultSetExtractor<List<RaffleReward>>() {
            @Override
            public List<RaffleReward> extractData(ResultSet rs) throws SQLException, DataAccessException {
                List<RaffleReward> _list = new ArrayList<>();

                while (rs.next()) {
                    RaffleReward reward = new RaffleReward();
                    reward.setTrr_id(rs.getInt("trr_id"));
                    reward.setRaffle_id(rs.getInt("raffle_id"));
                    reward.setReward_name(rs.getString("reward_name"));
                    reward.setReward_type(rs.getInt("reward_type"));
                    reward.setReward_picture(rs.getString("reward_picture"));
                    reward.setInventory(rs.getInt("inventory"));
                    reward.setHit_rate(rs.getDouble("hit_rate"));
                    reward.setReward_value(rs.getInt("reward_value"));
                    reward.setSort_id(rs.getInt("sort_id"));
                    _list.add(reward);
                }
                return _list;
            }
        });
        return list;
    }

    public int selectRaffleLogCount(final int memberId, final int raffleId, final boolean daily) {
        final String sql = "SELECT count(1) cnt FROM `T_RAFFLE_LOG` WHERE raffle_id=? AND member_id=?";

        StringBuilder sb = new StringBuilder(sql);
        if (daily) {
            long currS = dateTimeUtils.getMilliByToDay();
            LocalDate date = Instant.ofEpochMilli(currS).atZone(ZoneId.systemDefault()).toLocalDate();
            long currE = dateTimeUtils.getMilliByDateTime(date.getYear(), date.getMonthValue(), date.getDayOfMonth(), 23, 59, 59, 0);
            sb.append(" AND action_time>=").append(currS).append(" AND action_time<=").append(currE);
        }

        int result = 0;
        SqlRowSet rowSet = jdbcTemplate.queryForRowSet(sb.toString(), new Object[]{raffleId, memberId});
        if (rowSet.first()) {
            result = rowSet.getInt("cnt");
        }
        return result;
    }

    public RaffleReward selectRewardById(final int trrId) {
        final String sql ="SELECT trr_id,raffle_id,reward_name,reward_type,reward_picture,inventory,hit_rate,reward_value,sort_id " +
                "FROM `T_RAFFLE_REWARD` WHERE trr_id=?";

        RaffleReward reward = jdbcTemplate.query(sql, new Object[]{trrId}, new ResultSetExtractor<RaffleReward>() {
            @Override
            public RaffleReward extractData(ResultSet rs) throws SQLException, DataAccessException {
                RaffleReward _reward = null;

                if (rs.next()) {
                    _reward = new RaffleReward();
                    _reward.setTrr_id(rs.getInt("trr_id"));
                    _reward.setRaffle_id(rs.getInt("raffle_id"));
                    _reward.setReward_name(rs.getString("reward_name"));
                    _reward.setReward_type(rs.getInt("reward_type"));
                    _reward.setReward_picture(rs.getString("reward_picture"));
                    _reward.setInventory(rs.getInt("inventory"));
                    _reward.setHit_rate(rs.getInt("hit_rate"));
                    _reward.setReward_value(rs.getInt("reward_value"));
                    _reward.setSort_id(rs.getInt("sort_id"));
                }
                return _reward;
            }
        });
        return reward;
    }

    public int updateRewardInventory(final int trrId, final int inventory) {
        final String sql = "UPDATE `T_RAFFLE_REWARD` SET inventory=? WHERE trr_id=?";

        return jdbcTemplate.update(sql, inventory, trrId);
    }
}

File: member/manager-app/src/main/java/com/laf/manager/repository/RaffleRepository.kt
package com.laf.manager.repository

import com.laf.manager.dto.RaffleInfo
import com.laf.manager.dto.RaffleLog
import com.laf.manager.querycondition.raffle.RaffleFilterCondition
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.support.GeneratedKeyHolder
import org.springframework.stereotype.Repository
import java.sql.Date
import java.sql.Statement
import java.sql.Types

@Repository
class RaffleRepository {

    @Autowired
    lateinit var jdbcTemplate: JdbcTemplate

    fun insertRaffleInfo(raffleInfo: RaffleInfo): Int {
        val sql = "insert into `T_RAFFLE_INFO`(title,raffle_time_start,raffle_time_end,required_points," +
                "intro,mall_id,`times`,activity_code) " +
                "values(?,?,?,?,?,?,?,?)"

        val holder = GeneratedKeyHolder()

        jdbcTemplate.update({ conn ->
            val ps = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)
            ps.setString(1, raffleInfo.title)
            ps.setLong(2, raffleInfo.raffle_time_start)
            ps.setLong(3, raffleInfo.raffle_time_end)
            ps.setInt(4, raffleInfo.required_points)
            ps.setInt(5, raffleInfo.intro_id)
            ps.setInt(6, raffleInfo.mall_id)
            ps.setInt(7, raffleInfo.times)
            ps.setString(8, raffleInfo.activity_code)

            ps
        }, holder)

        return holder.key.toInt()
    }

    fun selectRaffleByCode(activity_code: String): List<RaffleInfo> {
        val sql = "select raffle_id, title,raffle_time_start,raffle_time_end,required_points," +
                "r.intro,a.content,r.mall_id,`times`,activity_code " +
                "from `T_RAFFLE_INFO` r inner join `T_ARTICLES` a on r.intro=a.article_id " +
                "where activity_code=?"

        return jdbcTemplate.query(sql, arrayOf(activity_code), {
            rs, _ ->
            val raffle = RaffleInfo(
                    rs.getInt("raffle_id"),
                    rs.getString("title"),
                    rs.getLong("raffle_time_start"),
                    rs.getLong("raffle_time_end"),
                    rs.getInt("required_points"),
                    rs.getInt("intro"),
                    rs.getString("content"),
                    rs.getInt("mall_id"),
                    rs.getInt("times"),
                    rs.getString("activity_code")
            )

            raffle
        })
    }

    fun selectRaffleById(raffleId: Int): List<RaffleInfo> {
        val sql = "select raffle_id, title,raffle_time_start,raffle_time_end,required_points," +
                "r.intro,a.content,r.mall_id,`times`,activity_code " +
                "from `T_RAFFLE_INFO` r inner join `T_ARTICLES` a on r.intro=a.article_id " +
                "where raffle_id=?"

        return jdbcTemplate.query(sql, arrayOf(raffleId), {
            rs, _ ->
            val raffle = RaffleInfo(
                    rs.getInt("raffle_id"),
                    rs.getString("title"),
                    rs.getLong("raffle_time_start"),
                    rs.getLong("raffle_time_end"),
                    rs.getInt("required_points"),
                    rs.getInt("intro"),
                    rs.getString("content"),
                    rs.getInt("mall_id"),
                    rs.getInt("times"),
                    rs.getString("activity_code")
            )

            raffle
        })
    }

    fun updateRaffleInfo(raffleInfo: RaffleInfo): Int {
        val sql = "update `T_RAFFLE_INFO` set title=?,raffle_time_start=?,raffle_time_end=?," +
                "required_points=?,times=?,activity_code=? where raffle_id=?"

        return jdbcTemplate.update(sql, raffleInfo.title, raffleInfo.raffle_time_start, raffleInfo.raffle_time_end,
                raffleInfo.required_points, raffleInfo.times,raffleInfo.activity_code, raffleInfo.raffle_id)
    }

    fun multipleSelectRaffleLogs(it: RaffleFilterCondition): List<RaffleLog> {
        var sql = "SELECT i.title,l.member_name,l.member_mobile,i.required_points,l.trr_id,l.action_time,r.reward_name,r.reward_type " +
                "FROM `T_RAFFLE_LOG` l INNER JOIN t_raffle_info i ON l.raffle_id=i.raffle_id LEFT JOIN t_raffle_reward r ON l.trr_id=r.trr_id " +
                "WHERE 1=1"

        val args = arrayListOf<Any>()
//        var argTypes = arrayListOf<Int>()
//        argTypes.add(Types.INTEGER)

        if (it.title.isNotEmpty()) {
            sql += " AND i.title LIKE '%' ? '%'"
            args.add(it.title)
        }

        if (it.username.isNotEmpty()) {
            sql += " AND l.member_name LIKE '%' ? '%'"
            args.add(it.username)
        }

        if (it.mobile.isNotEmpty()) {
            sql += " AND l.member_mobile=?"
            args.add(it.mobile)
        }

        if (it.actionTimeStart > 0 && it.actionTimeEnd > 0) {
            sql += " AND l.action_time>=? AND l.action_time<=?"
            args.add(it.actionTimeStart)
            args.add(it.actionTimeEnd)
        }

        if (it.isWin != -1) {
            if (it.isWin == 1) {
                sql += " AND l.trr_id>0"
            } else if (it.isWin == 0) {
                sql += " AND l.trr_id=0"
            }
        }

        sql += " ORDER BY l.action_time DESC LIMIT ?,?"
        args.add(it.offset)
        args.add(it.size)
//        argTypes.add(Types.INTEGER)
//        argTypes.add(Types.INTEGER)

        val arr = args.toArray()

        return jdbcTemplate.query(sql, arr, {
            rs, _ ->
            val log = RaffleLog(
                    rs.getString("title"),
                    rs.getString("member_name"),
                    rs.getString("member_mobile"),
                    Date(rs.getLong("action_time")),
                    rs.getInt("required_points"),
                    rs.getInt("trr_id"),
                    rs.getString("reward_name"),
                    rs.getInt("reward_type")
            )
            log
        })
    }

    fun multipleSelectRaffleLogsCount(it: RaffleFilterCondition): Int {
        var sql = "SELECT count(1) cnt " +
                "FROM `T_RAFFLE_LOG` l INNER JOIN t_raffle_info i ON l.raffle_id=i.raffle_id LEFT JOIN t_raffle_reward r ON l.trr_id=r.trr_id " +
                "WHERE 1=1"

        val args = arrayListOf<Any>()
        var argTypes = arrayListOf<Int>()

        if (it.title.isNotEmpty()) {
            sql += " AND i.title LIKE '%' ? '%'"
            args.add(it.title)
            argTypes.add(Types.VARCHAR)
        }

        if (it.username.isNotEmpty()) {
            sql += " AND l.member_name LIKE '%' ? '%'"
            args.add(it.username)
            argTypes.add(Types.VARCHAR)
        }

        if (it.mobile.isNotEmpty()) {
            sql += " AND l.member_mobile=?"
            args.add(it.mobile)
            argTypes.add(Types.VARCHAR)
        }

        if (it.actionTimeStart > 0 && it.actionTimeEnd > 0) {
            sql += " AND l.action_time>=? AND l.action_time<=?"
            args.add(it.actionTimeStart)
            args.add(it.actionTimeEnd)
            argTypes.add(Types.BIGINT)
            argTypes.add(Types.BIGINT)
        }

        if (it.isWin != -1) {
            if (it.isWin == 1) {
                sql += " AND l.trr_id>0"
            } else if (it.isWin == 0) {
                sql += " AND l.trr_id=0"
            }
        }

        val arr = args.toArray()
        val artype = argTypes.toIntArray()
//        val artype = arrayOfNulls<Int>(arr.size)
//        argTypes.toArray(artype)

        val rowSet = jdbcTemplate.queryForRowSet(sql, arr, artype)

        var count = 0
        if (rowSet.first()) {
            count = rowSet.getInt("cnt")
        }
        return count
    }

    /*
    * public List<Member> multipleSelectMembers(final MemberFilterCondition $$) {
        String sql = "select m.member_id,m.name,m.mobile,m.sex,m.birthday,m.occupation,m.address," +
                "m.degree_of_education,m.income_range,m.interest,m.wechat_account,m.enable_public_wa," +
                "m.cumulate_points,m.usable_points,m.cumulate_amount,m.level,m.level_id,mm.open_id,mm.status," +
                "m.member_card_no,mm.regist_date,mm.mall_id,m.email,m.edit_flag,m.birthday_modified " +
                "from `T_MEMBER` m inner join `T_MALL_MEMBER` mm on m.member_id=mm.member_id " +
                "where mm.mall_id=?";

        List<Object> args = new ArrayList<>();
        args.add($$.getMallId());

        if (!StringUtils.isEmpty($$.getUsername())) {
            sql += " and m.name like '%' ? '%'";
            args.add($$.getUsername());
        }

        if (!StringUtils.isEmpty($$.getMobile())) {
            sql += " and m.mobile = ?";
            args.add($$.getMobile());
        }

        if ($$.getRegisterDateStart() > 0L && $$.getRegisterDateEnd() > 0L) {
            sql += " and mm.regist_date >= ? and mm.regist_date<=?";
            args.add($$.getRegisterDateStart());
            args.add($$.getRegisterDateEnd());
        }

        if ($$.getCumulateAmountStart().doubleValue() > 0 && $$.getCumulateAmountEnd().doubleValue() > 0) {
            sql += " and m.cumulate_amount>=? and m.cumulate_amount<=?";
            args.add($$.getCumulateAmountStart());
            args.add($$.getCumulateAmountEnd());
        }

        if ($$.getCumulatePointsStart() > 0 && $$.getCumulatePointsEnd() > 0) {
            sql += " and m.cumulate_points>=? and m.cumulate_points<=?";
            args.add($$.getCumulatePointsStart());
            args.add($$.getCumulatePointsEnd());
        }

        if ($$.getLevel() > 0) {
            sql += " and m.level_id=?";
            args.add($$.getLevel());
        }

        if ($$.getOccupation() >= 0) {
            sql += " and m.occupation=?";
            args.add($$.getOccupation());
        }

        if ($$.getBirthdayStart() > 0L && $$.getBirthdayEnd() > 0L) {
            sql += " and m.birthday>=? and m.birthday<=?";
            args.add($$.getBirthdayStart());
            args.add($$.getBirthdayEnd());
        }

        if ($$.getStatus() > -1) {
            sql += " and mm.status=?";
            args.add($$.getStatus());
        }

        sql += " order by mm.regist_date desc limit ?, ?";
        args.add($$.getOffset());
        args.add($$.getSize());

        return jdbcTemplate.query(sql, args.toArray(),
                (rs, rowNum) -> {
                    Member $ = new Member();
                    $.setMember_id(rs.getInt("member_id"));
                    $.setName(rs.getString("name"));
                    $.setMobile(rs.getString("mobile"));
                    $.setSex(rs.getInt("sex"));
                    $.setBirthday(new Date(rs.getLong("birthday")));
                    $.setOccupation(rs.getInt("occupation"));
                    $.setAddress(rs.getString("address"));
                    $.setDegree_of_education(rs.getInt("degree_of_education"));
                    $.setIncome_range(rs.getInt("income_range"));
                    $.setInterest(rs.getString("interest"));
                    $.setWechat_account(rs.getString("wechat_account"));
                    $.setEnable_public_wa(rs.getBoolean("enable_public_wa"));
                    $.setCumulate_points(rs.getInt("cumulate_points"));
                    $.setUsable_points(rs.getInt("usable_points"));
                    $.setMall_id(rs.getInt("mall_id"));
                    $.setCumulate_amount(rs.getBigDecimal("cumulate_amount"));
                    $.setLevel(rs.getString("level"));
                    $.setLevel_id(rs.getInt("level_id"));
                    $.setMember_card_no(rs.getString("member_card_no"));
                    $.setRegist_date(new Date(rs.getLong("regist_date")));
                    $.setOpen_id(rs.getString("open_id"));
                    $.setEmail(rs.getString("email"));
                    $.setEdit_flag(rs.getInt("edit_flag"));
                    $.setBirthday_modified(rs.getInt("birthday_modified"));
                    $.setStatus(rs.getInt("status"));
                    return $;
                });
    }
    * */

}

